### PR TITLE
Jetpack DNA: Add full classmap support to Autoloader

### PR DESCRIPTION
**Pull request by mikejolley**
This PR tweaks the Jetpack-Autoloader package to offer greater support for classmaps.

Currently, the "legacy" classmap support simply uses version `0.1` when registering the classes. That means it's totally random which class would be autoloaded should the same class exist in multiple places.

This creates classmap generation in the same way it generates for psr-4 - versioned, with `exclude-from-classmap` support

I'm testing this branch with WooCommerce + the WooCommerce REST API package which includes a classmap for it's legacy classes which are not psr-4. Example rules:

```
	'WC_REST_Tax_Classes_V1_Controller' => array(
		'version' => 'dev-test/jetpack-autoloader',
		'path'    => $vendorDir . '/woocommerce-rest-api/src/Controllers/Version1/class-wc-rest-tax-classes-v1-controller.php'
	),
	'WC_REST_Webhook_Deliveries_V1_Controller' => array(
		'version' => 'dev-test/jetpack-autoloader',
		'path'    => $vendorDir . '/woocommerce-rest-api/src/Controllers/Version1/class-wc-rest-webhook-deliveries-v1-controller.php'
	),
```

I'm not sure what process you all use to get this code into the `jetpack-autoloader` package itself which is one of the dependencies I'm using, but please let me know how I can get this shipped or if you spot any issues :)

**Testing instructions:**

This is not something you can test in the UI. You need to edit `composer.json` and add a `classmap` entry in the autoload section e.g. 

```
"autoload": {
    "classmap": [
      "includes"
    ],
    "exclude-from-classmap": [
      "includes/legacy",
      "includes/libraries"
    ]
  },
```

Then you'd need to regenerate the autoload files `composer dump-autoload` and check `vendor/composer/autoload_classmap_package.php` to see if it has the classes in your classmap definition.

**Proposed changelog entry for your changes:**

> Added classmap and exclude-from-classmap support to the jetpack-autoloader package.

### CODE REVIEW

**Comment by brbrr** [comment]
> I'm not sure what process you all use

few bits of that process is mentioned by @jetpackbot above. Adding descriptive testing instructions, and proposed changelog entry _will_ help us review and test your changes

Also, adding some unit tests for your changes will also help. 

As per https://github.com/Automattic/vaultpress/blob/master/composer.json#L14, you can use the `master`s version of the package directly. Once this PR is landed - it should be available to use.

File: packages/autoloader/src/AutoloadGenerator.php [hunk 1]
```diff
@@ -143,10 +163,15 @@ protected function parseAutoloadsType( array $packageMap, $type, PackageInterfac
 	 * @return array $classMap
 	 */
 	private function getClassMap( array $autoloads, Filesystem $filesystem, $vendorPath, $basePath ) {
-		$blacklist      = null; // not supported for now.
+		$blacklist = null;
+
+        if ( ! empty( $autoloads['exclude-from-classmap'] ) ) {
+            $blacklist = '{(' . implode( '|', $autoloads['exclude-from-classmap'] ) . ')}';
```
Comment (lezama) [inline packages/autoloader/src/AutoloadGenerator.php]: seems like we are using spaces instead of tabs here

**Review by lezama — DISMISSED** [review]
nice work!
I tested this and it seems produce a valid and working class map.

let's merge this and generate the next beta with it @jeherve ?

File: packages/autoloader/src/AutoloadGenerator.php [hunk 2]
```diff
@@ -111,24 +118,37 @@ protected function parseAutoloadsType( array $packageMap, $type, PackageInterfac
...
 			}
-			foreach ( $autoload['psr-4'] as $namespace => $paths ) {
-				foreach ( (array) $paths as $path ) {
-					$relativePath              = empty( $installPath ) ? ( empty( $path ) ? '.' : $path ) : $installPath . '/' . $path;
-					$autoloads[ $namespace ][] = array(
-						'path'    => $relativePath,
-						'version' => $package->getVersion(), // Version of the class comes from the package - should we try to parse it?
-					);
+
+			if ( 'psr-4' == $type && isset( $autoload['psr-4'] ) && is_array( $autoload['psr-4'] ) ) {
```
Comment (lezama) [inline packages/autoloader/src/AutoloadGenerator.php]: let's use strict comparison

File: packages/autoloader/src/AutoloadGenerator.php [hunk 3]
```diff
@@ -111,24 +118,37 @@ protected function parseAutoloadsType( array $packageMap, $type, PackageInterfac
...
+						$relativePath              = empty( $installPath ) ? ( empty( $path ) ? '.' : $path ) : $installPath . '/' . $path;
+						$autoloads[ $namespace ][] = array(
+							'path'    => $relativePath,
+							'version' => $package->getVersion(), // Version of the class comes from the package - should we try to parse it?
+						);
+					}
+				}
+			}
+
+			if ( 'classmap' == $type && isset( $autoload['classmap'] ) && is_array( $autoload['classmap'] ) ) {
```
Comment (lezama) [inline packages/autoloader/src/AutoloadGenerator.php]:
```suggestion
			if ( 'classmap' === $type && isset( $autoload['classmap'] ) && is_array( $autoload['classmap'] ) ) {
```

**Review by lezama — APPROVED** [review]

**Comment by zinigor** [comment]
Ported into `branch-7.5` in a7574633a.

**Pull request merged**